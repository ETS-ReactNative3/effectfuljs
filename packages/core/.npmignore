node_modules
src
test
.tern*
.babelrc
.DS_Store
.git
dist
tmp-*
*.orig
*.log
*~
*.map
npm-debug.log

