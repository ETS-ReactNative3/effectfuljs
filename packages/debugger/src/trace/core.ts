import config from "../config";
import { saved } from "../state";

export interface Record {
  undone: boolean;
  version: number;
  operations?: Operation;
  prev: Record | null;
  next: Record | null;
}

/** operations trace - callbacks to be executed on reverting state */
export interface Operation {
  prev?: Operation;
  (): void;
}

/** extension point for tracing customization */
export const traceMetaSymbol = Symbol("@effectful/debugger/traceMeta");

/** trace related information stored in the object */
export const traceDataSymbol = Symbol("@effectful/debugger/traceData");

export interface MaybeTarget<T = unknown> {
  [traceMetaSymbol]?: TraceMeta<T>;
  [traceDataSymbol]?: TraceData<T>;
}

export interface Target<T = unknown> {
  [traceMetaSymbol]: TraceMeta<T>;
  [traceDataSymbol]: TraceData<T>;
}

export interface TraceMeta<T = unknown> {
  wrap(target: T): T;
  proto?: object | null;
}

export const notTraceable: TraceMeta<any> = {
  wrap(target: any): any {
    target[traceDataSymbol] = false;
    return target;
  }
};

/** Class for representing trace data steps */
export class TraceData<T> {
  meta: TraceMeta<T>;
  proxy: T | null;
  target: T;
  version: number;
  lastUpdate: Record | null;
  prototypeChanged: boolean;
  proto?: any;

  constructor(meta: TraceMeta<T>, proxy: any, target: any) {
    this.meta = meta;
    this.proxy = proxy;
    this.target = target;
    this.lastUpdate = journal.trace;
    this.version = journal.trace ? journal.trace.version : 0;
    this.prototypeChanged = false;
  }
}

interface Journal {
  trace: Record | null;
  paused: boolean;
}

export const journal: Journal = {
  /** a list of changes */
  trace: null,
  /** this is true when `undoing` changes to avoid re-recording them */
  paused: !config.timeTravel
};

/**
 * Returns an object which can be passed
 * to `undo` to restored the current state
 */
export function checkpoint(): Record {
  const next = {
    undone: false,
    version: journal.trace ? journal.trace.version + 1 : 0,
    prev: journal.trace,
    next: null
  };
  if (journal.trace) journal.trace.next = next;
  journal.trace = next;
  return next;
}

/**
 * Registers a function(`f`) to be executed on `undo` call
 * The functions are executed in backward order
 */
export function record(f: Operation) {
  if (!f) return;
  if (journal.trace && !journal.paused) {
    f.prev = journal.trace.operations || undefined;
    journal.trace.operations = f;
  }
}

export function undoImpl(from?: Record): Record | null {
  const { trace } = journal;
  if (!trace) return null;
  if (!from) from = trace;
  /*
  for (let i: Record | null = from; i != null; i = i.next) {
    const destVersion = i.version;
    if (i.undone) throw new TypeError("The checkpoint was already undone");
    i.undone = true;
    for (let j = i.snapshots; j; j = j.prev) {
      if (j.data.version >= destVersion) {
        j.call();
        j.data.version = destVersion;
      }
    }
    trace.snapshots = void 0;
  }
    */
  for (
    let i: Record | null = trace, till = from.prev;
    i != null && i != till;
    i = i.prev
  )
    for (let j = i.operations; j != null; j = j.prev) j();
  journal.trace = from.prev;
  if (from.prev) from.prev.next = null;
  return from;
}

/** Returns own property names and symbols */
export function* allProps(obj: any): Iterable<string | symbol> {
  yield* Object.getOwnPropertyNames(obj);
  yield* Object.getOwnPropertySymbols(obj);
}

export function captureSnapshot<T>(data: TraceData<T>) {
  if (journal.paused || !journal.trace || data.lastUpdate === journal.trace)
    return;
  const target = data.target;
  const lastUpdate = data.lastUpdate;
  data.lastUpdate = journal.trace;
  data.prototypeChanged = false;
  const snapshot = Object.getOwnPropertyDescriptors(target);
  record(() => {
    data.lastUpdate = lastUpdate;
    if (data.prototypeChanged) Object.setPrototypeOf(target, data.proto);
    for (const i of allProps(target))
      if (!(i in snapshot)) delete (<any>target)[i];
    for (const i of allProps(snapshot)) {
      const descr = (<any>snapshot)[i];
      if (!descr.set) {
        if (descr.enumberable && descr.writeable)
          (<any>target)[i] = descr.value;
        else Object.defineProperty(target, i, descr);
      }
    }
  });
}

export const defaultTraps: ProxyHandler<any> = {
  set(target, prop, value) {
    captureSnapshot(target[traceDataSymbol]);
    return Reflect.set(target, prop, value);
  },
  setPrototypeOf(target, proto) {
    const data = target[traceDataSymbol];
    captureSnapshot(data);
    data.prototypeChanged = true;
    data.proto = Object.getPrototypeOf(target);
    return Reflect.setPrototypeOf(target, unwrapTarget(proto));
  },
  // TODO: getters of __proto__ property should return wrapped prototype too
  getPrototypeOf(target) {
    return getWrapper(Object.getPrototypeOf(target));
  },
  deleteProperty(target, prop) {
    captureSnapshot(target[traceDataSymbol]);
    return Reflect.deleteProperty(target, prop);
  }
};

export const defaultTrapsWithGetter: ProxyHandler<any> = {
  ...defaultTraps,
  get(target, prop) {
    return Reflect.get(target, prop);
  }
};

export function defaultWrap<T extends MaybeTarget<T>>(
  meta: TraceMeta<T>,
  target: T,
  traps: ProxyHandler<T> = defaultTraps
): T {
  const { proto } = meta;
  if (proto) {
    const origProto = Object.getPrototypeOf(target);
    traps = {
      ...traps,
      getPrototypeOf() {
        return origProto;
      }
    };
    traps.getPrototypeOf = () => origProto;
    Object.setPrototypeOf(target, proto);
  }
  const proxy = new saved.Proxy(target, traps);
  target[traceDataSymbol] = new TraceData(meta, proxy, target);
  return proxy;
}

export const defaultTraceMeta: TraceMeta<any> = {
  wrap(arg: any) {
    return defaultWrap(this, arg);
  },
  proto: null
};

export function wrapImpl<T extends MaybeTarget<T>>(target: T): T {
  if (!target) return target;
  const type = typeof target;
  if (type === "object" || type === "function") {
    if (target.hasOwnProperty(traceDataSymbol)) return target;
    return (target[traceMetaSymbol] || defaultTraceMeta).wrap(target);
  }
  return target;
}

/**
 * if `target` is a proxy returns its enclosing value
 * or returns `target` otherwise
 */
export function unwrapTarget<T extends MaybeTarget<T>>(target: T): T {
  if (!target || !target.hasOwnProperty(traceDataSymbol)) return target;
  return (<any>target)[traceDataSymbol].target;
}

/**
 * for values tracked by some proxy returns the proxy,
 * or returns `target` if it is already the proxy
 */
export function getWrapper<T extends MaybeTarget<T>>(maybeProxy: T): T {
  if (!maybeProxy || !maybeProxy.hasOwnProperty(traceDataSymbol))
    return maybeProxy;
  return (<any>maybeProxy)[traceDataSymbol].proxy;
}
