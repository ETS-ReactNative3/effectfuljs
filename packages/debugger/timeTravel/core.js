"use strict";

exports.__esModule = true;
exports.reset = reset;
exports.checkpoint = checkpoint;
exports.replay = replay;
exports.allProps = allProps;
exports.captureSnapshot = captureSnapshot;
exports.defaultWrap = defaultWrap;
exports.wrapImpl = wrapImpl;
exports.unwrapTarget = unwrapTarget;
exports.getWrapper = getWrapper;
exports.defaultTraceMeta = exports.defaultTrapsWithGetter = exports.defaultTraps = exports.record = exports.journal = exports.TraceData = exports.notTraceable = exports.traceDataSymbol = exports.traceMetaSymbol = void 0;

var _config = require("../config");

var _state = require("../state");

var _serialization = require("@effectful/serialization");

/** extension point for tracing customization */
const traceMetaSymbol = Symbol("@effectful/debugger/traceMeta");
/** trace related information stored in the object */

exports.traceMetaSymbol = traceMetaSymbol;
const traceDataSymbol = Symbol("@effectful/debugger/traceData");
exports.traceDataSymbol = traceDataSymbol;
const notTraceable = {
  wrap(target) {
    target[traceDataSymbol] = false;
    return target;
  }

};
/** Class for representing trace data steps */

exports.notTraceable = notTraceable;

class TraceData {
  constructor(meta, proxy, target) {
    this.meta = meta;
    this.proxy = proxy;
    this.target = target;
    this.lastUpdate = journal.now;
    this.prototypeChanged = false;
  }

}

exports.TraceData = TraceData;
const journal = {
  now: null,
  past: null,
  future: null
};
exports.journal = journal;

function reset() {
  journal.past = journal.now = journal.future = null;
}
/**
 * Returns an object which can be passed
 * to `undo` to restored the current state
 */


function checkpoint() {
  journal.future = null;
  const {
    now
  } = journal;
  if (now) now.prev = journal.past;
  journal.past = now;
  return journal.now = {
    prev: null
  };
}
/**
 * Registers a function(`f`) to be executed on replay in backward order
 */


const record = _config.default.timeTravel ? function record(f) {
  if (!f) return;
  const {
    now
  } = journal;
  if (!now) return;
  f.prev = now.operations || undefined;
  now.operations = f;
} : function () {};
/** runs all actions stored in the cocord */

exports.record = record;

function replay(rec) {
  for (let i = rec.operations; i != null; i = i.prev) i.call();
}
/** Returns own property names and symbols */


function* allProps(obj) {
  yield* Object.getOwnPropertyNames(obj);
  yield* Object.getOwnPropertySymbols(obj);
}

class ObjectSnapshot {
  constructor(data) {
    this.data = data;
    this.at = data.lastUpdate;
    this.snapshot = Object.getOwnPropertyDescriptors(data.target);
    data.lastUpdate = journal.now;
    data.prototypeChanged = false;
  }

  call() {
    const {
      data,
      at,
      snapshot
    } = this;
    const {
      prototypeChanged,
      proto,
      target
    } = data;

    if (_config.default.timeTravelForward) {
      data.lastUpdate = null;
      captureSnapshot(data);

      if (prototypeChanged) {
        data.proto = Object.getPrototypeOf(target);
        data.prototypeChanged = true;
      }
    }

    data.lastUpdate = at;
    if (prototypeChanged) Object.setPrototypeOf(target, proto);

    for (const i of allProps(target)) if (!(i in snapshot)) delete target[i];

    for (const i of allProps(snapshot)) {
      const descr = snapshot[i];

      if (!descr.set) {
        if (descr.enumberable && descr.writeable) target[i] = descr.value;else Object.defineProperty(target, i, descr);
      }
    }
  }

}

(0, _serialization.regConstructor)(ObjectSnapshot);

function captureSnapshot(data) {
  if (!journal.now || data.lastUpdate === journal.now) return;
  record(new ObjectSnapshot(data));
}

const defaultTraps = {
  set(target, prop, value) {
    captureSnapshot(target[traceDataSymbol]);
    return Reflect.set(target, prop, value);
  },

  setPrototypeOf(target, proto) {
    const data = target[traceDataSymbol];
    captureSnapshot(data);
    data.prototypeChanged = true;
    data.proto = Object.getPrototypeOf(target);
    return Reflect.setPrototypeOf(target, unwrapTarget(proto));
  },

  // TODO: getters of __proto__ property should return wrapped prototype too
  getPrototypeOf(target) {
    return getWrapper(Object.getPrototypeOf(target));
  },

  deleteProperty(target, prop) {
    captureSnapshot(target[traceDataSymbol]);
    return Reflect.deleteProperty(target, prop);
  }

};
exports.defaultTraps = defaultTraps;
const defaultTrapsWithGetter = { ...defaultTraps,

  get(target, prop) {
    return Reflect.get(target, prop);
  }

};
exports.defaultTrapsWithGetter = defaultTrapsWithGetter;

function defaultWrap(meta, target, traps = defaultTraps) {
  const {
    proto
  } = meta;

  if (proto) {
    const origProto = Object.getPrototypeOf(target);
    traps = { ...traps,

      getPrototypeOf() {
        return origProto;
      }

    };

    traps.getPrototypeOf = () => origProto;

    Object.setPrototypeOf(target, proto);
  }

  const proxy = new _state.saved.Proxy(target, traps);
  target[traceDataSymbol] = new TraceData(meta, proxy, target);
  return proxy;
}

const defaultTraceMeta = {
  wrap(arg) {
    return defaultWrap(this, arg);
  },

  proto: null
};
exports.defaultTraceMeta = defaultTraceMeta;

function wrapImpl(target) {
  if (!target) return target;
  const type = typeof target;

  if (type === "object" || type === "function") {
    if (target.hasOwnProperty(traceDataSymbol)) return target;
    return (target[traceMetaSymbol] || defaultTraceMeta).wrap(target);
  }

  return target;
}
/**
 * if `target` is a proxy returns its enclosing value
 * or returns `target` otherwise
 */


function unwrapTarget(target) {
  if (!target || !target.hasOwnProperty(traceDataSymbol)) return target;
  return target[traceDataSymbol].target;
}
/**
 * for values tracked by some proxy returns the proxy,
 * or returns `target` if it is already the proxy
 */


function getWrapper(maybeProxy) {
  if (!maybeProxy || !maybeProxy.hasOwnProperty(traceDataSymbol)) return maybeProxy;
  return maybeProxy[traceDataSymbol].proxy;
}