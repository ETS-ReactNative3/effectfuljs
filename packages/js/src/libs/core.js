import * as Kit from "../kit"
import * as Policy from "../policy"
import * as Debug from "../debug"

const fullOpts = {
  bindCalls: {
    byNs: {
      console:false,
      process:false,
      pack:false
    },
    libNs: {
      reify: false
    },
    all: true,
    ns: true
  },
  generator:false,
  coerce:true,
  transform:true
}

const minOpts = {
  bindCalls: {
    ns: true,
    libNs: {
      reflect: true
    }
  },
  coerce:true,
  generator:false,
  transform:true
}

const disabledOpts = {
  transform: false
}

const generatorOps = {
  YieldExpression: true,
  AwaitExpression: true
}

const postproc = Kit.pipe(
  Policy.setFuncOpts(disabledOpts),
  Policy.profiles,
  Policy.setQNames)

export default function* coreInit() {
  const generators = Policy.injectFuncOpts({
    generator:true,
    async:false,
    esRebind:true,
    static:true,
    transform:true,
    ops:generatorOps,
    coerce:false,
    scopePrefix:true,
    scopeConstructor:"generator",
    bindName:"yldStar",
    requireFinalPure:true,
    pureForOf:true
  })
  const asyncDo = Policy.injectFuncOpts({
    generator:false,
    async:true,
    bindCalls: {},
    esRebind:true,
    ops:generatorOps,
    coerce:false,
    transform:true,
    scopePrefix:true,
    scopeConstructor:"async",
    bindName:"chain",
    requireFinalPure:false,
    pureForOf:true
  })
  const asyncGeneratorsDo = Policy.injectFuncOpts({
    generator:true,
    async:true,
    bindCalls: {},
    esRebind:true,
    ops: generatorOps,
    coerce:false,
    transform:true,
    scopePrefix:true,
    scopeConstructor:"asyncGenerator",
    bindName:"chain",
    requireFinalPure:false,
    pureForOf:true
  })
  const disabled = Policy.injectFuncOpts(disabledOpts)
  return Kit.pipe(
    function*(si) {
      const s = Kit.auto(si)
      yield Kit.tok(Policy.configDiff,{
        node: {
          profiles: {
            disabled,
            full:Kit.pipe(Policy.injectOpts(fullOpts),
                        Policy.injectFuncOpts(fullOpts)),
            defaultFull:Policy.injectFuncOpts(fullOpts),
            minimal:Kit.pipe(Policy.injectOpts(minOpts),
                           Policy.injectFuncOpts(minOpts)),
            defaultMinimal:Policy.injectFuncOpts(minOpts),
            generators,asyncDo,asyncGeneratorsDo,
            es:function* () {
              yield* disabled()
              yield* generators()
              yield* asyncGeneratorsDo()
              yield* asyncDo()
            }
          }
        }
      })
      let {profile} = s.opts
      if (profile != null) {
        if (!Array.isArray(profile))
          profile = [profile]
        for(const i of profile)
          yield Kit.tok(Policy.profile,{node:{name:i}})
      }
      yield* s
    },
    postproc)
}

