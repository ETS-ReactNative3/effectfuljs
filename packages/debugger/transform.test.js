import * as babel from "@babel/core";
import * as path from "path";

test("default transpiler", function() {
  expect(
    babel.transformFileSync("__fixtures__/code.js").code
  ).toMatchSnapshot();
  expect(
    babel.transformFileSync("__fixtures__/mod1.js").code
  ).toMatchSnapshot();
  expect(
    babel.transformFileSync("__fixtures__/mod2.js").code
  ).toMatchSnapshot();
});
