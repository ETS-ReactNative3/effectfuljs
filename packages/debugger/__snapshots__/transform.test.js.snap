// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default transpiler 1`] = `
"\\"use strict\\";

var M = require(\\"../main\\");

var $module = M.module(\\"/__fixtures__/code.js\\"),
    meta$f3 = M.meta($module, {}),
    meta$objFn = M.meta($module, {}),
    meta$fn1 = M.meta($module, {
  v5: [\\"v5\\", \\"41:8-41:10\\", \\"40:29-43:1\\"]
}),
    meta$func = M.meta($module, {}),
    meta$f2 = M.meta($module, {}),
    meta$funcAsync = M.meta($module, {}),
    meta$f1 = M.meta($module, {}),
    _meta$f = M.meta($module),
    meta$f = M.meta($module, {}),
    meta$C = M.meta($module, {
  v1: [\\"v1\\", \\"6:6-6:8\\", \\"5:14-19:1\\"],
  v2: [\\"v2\\", \\"8:10-8:12\\", \\"7:6-16:3\\"],
  v3: [\\"v3\\", \\"9:8-9:10\\", \\"5:0-19:1\\"],
  i: [\\"i\\", \\"11:15-11:16\\", \\"11:4-13:5\\"],
  i1: [\\"i\\", \\"14:13-14:14\\", \\"5:0-19:1\\"],
  v4: [\\"v4\\", \\"15:8-15:10\\", \\"5:0-19:1\\"],
  e: [\\"e\\", \\"16:11-16:12\\", \\"5:14-19:1\\"]
});

module.exports = M.fun(function f(module, exports) {
  var _Cl;

  var ctx = M.instance(f);
  ctx.$.C = M.fun(function _C(a) {
    var C = M.instance(_C);
    C.$$.f = ctx.$;
    C.$.a = a;
    return C.scope(5);
  }, meta$C, C_1, C_err, C_fin);
  ctx.$.Cl = (_Cl = class Cl {
    constructor() {
      var v4, f;
      not_traced(f.v0, v4);
    }

    get prop1() {
      not_traced();
    }

    set prop1(value) {
      not_traced();
    }

    funcAsync() {
      var _ctx = M.instance(_Cl.prototype.funcAsync, meta$funcAsync, funcAsync_1);

      _ctx.$$.f = ctx.$;
      return _ctx.scope(5);
    }

    func() {
      var ctx = M.instance(_Cl.prototype.func, meta$func, func_1);
      return ctx.scope(5);
    }

  }, M.fun(_Cl.prototype.funcAsync, meta$funcAsync, funcAsync_1), M.fun(_Cl.prototype.func, meta$func, func_1), _Cl);
  ctx.$.module = module;
  ctx.$.exports = exports;
  return ctx.scope(5);
}, meta$f3, _f_1)({}, {});

function C_err(s) {
  switch (s) {
    case 7:
    case 8:
    case 9:
    case 10:
    case 11:
    case 12:
    case 13:
    case 17:
    case 18:
    case 22:
    case 23:
      return 28;

    case 14:
    case 15:
      return 29;

    case 16:
      return 31;

    case 19:
    case 20:
      return 32;

    case 21:
      return 34;

    default:
      return 1;
  }
}

function C_fin(s) {
  switch (s) {
    case 14:
    case 15:
      return 30;

    case 19:
    case 20:
      return 33;

    default:
      return 0;
  }
}

function C_1(s, p) {
  var C = this;

  for (;;) {
    try {
      switch (s) {
        case 5:
          return C.chain(M.brk(\\"statement\\", \\"6:2-6:14\\"), 6);

        case 6:
          C.$.v1 = 10;
          return C.chain(M.brk(\\"statement\\", \\"7:2-18:3\\"), 7);

        case 7:
          return C.chain(M.brk(\\"statement\\", \\"8:4-8:18\\"), 8);

        case 8:
          C.$.v2 = 11;
          return C.chain(M.brk(\\"statement\\", \\"9:4-9:11\\"), 9);

        case 9:
          return C.chain(M.brk(\\"statement\\", \\"10:4-10:42\\"), 10);

        case 10:
          return C.chain(C.$$.f.mod.fn1(C.$.a), 11);

        case 11:
          return C.chain(C.$$.f.mod.proc1(p, C.$$.f.v0, C.$.v1, C.$.v2, C.$.v3), 12);

        case 12:
          return C.chain(M.brk(\\"statement\\", \\"11:4-13:5\\"), 13);

        case 13:
          C.$.loop = M.iterator(something);
          s = 14;
          continue;

        case 14:
          if (!(C.$.i2 = C.$.loop.next()).done) {
            C.$.i = C.$.i2.value;
            return C.chain(M.brk(\\"statement\\", \\"12:6-12:19\\"), 15);
          } else {
            C.$.fc = 17;
            s = 16;
            continue;
          }

        case 15:
          return C.chain(C.$$.f.mod.proc2(C.$.i), 14);

        case 16:
          if (C.$.loop.return) {
            C.$.loop.return();
          }

          C.$.ex = C.$.err1;
          s = C.$.fc;
          continue;

        case 17:
          return C.chain(M.brk(\\"statement\\", \\"14:4-14:42\\"), 18);

        case 18:
          C.$.loop1 = M.iterator(something);
          s = 19;
          continue;

        case 19:
          if (!(C.$.i3 = C.$.loop1.next()).done) {
            C.$.i1 = C.$.i3.value;
            return C.chain(M.brk(\\"statement\\", \\"14:29-14:42\\"), 20);
          } else {
            C.$.fc1 = 22;
            s = 21;
            continue;
          }

        case 20:
          return C.chain(C.$$.f.mod.proc3(C.$.i1), 19);

        case 21:
          if (C.$.loop1.return) {
            C.$.loop1.return();
          }

          C.$.ex = C.$.err2;
          s = C.$.fc1;
          continue;

        case 22:
          return C.chain(M.brk(\\"statement\\", \\"15:4-15:17\\"), 23);

        case 23:
          C.$.v4 = 100;
          s = 26;
          continue;

        case 24:
          C.$.e = C.$.ex;
          return C.chain(M.brk(\\"statement\\", \\"17:4-17:24\\"), 25);

        case 25:
          return C.chain(C.$$.f.mod.proc4(C.$.e, C.$.v4, C.$.i1), 26);

        case 26:
          return C.pure();

        case 0:
          return C.pure(p);

        case 1:
          throw p;

        case 27:
          s = 0;
          p = C.$.r;
          continue;

        case 28:
          C.$.ex = p;
          s = 24;
          continue;

        case 29:
          C.$.fc = 24, C.$.err1 = p;
          s = 16;
          continue;

        case 30:
          C.$.fc = 27, C.$.r = p;
          s = 16;
          continue;

        case 31:
          C.$.ex = p;
          s = 24;
          continue;

        case 32:
          C.$.fc1 = 24, C.$.err2 = p;
          s = 21;
          continue;

        case 33:
          C.$.fc1 = 27, C.$.r = p;
          s = 21;
          continue;

        case 34:
          C.$.ex = p;
          s = 24;
          continue;

        default:
          throw new Error(\\"invalid state\\");
      }
    } catch (ex) {
      if ((p = ex, s = C_err(s)) !== 1) continue;
      throw ex;
    }
  }
}

function funcAsync_1(s, p) {
  var _ctx = this;

  switch (s) {
    case 5:
      return _ctx.chain(M.brk(\\"statement\\", \\"33:4-33:19\\"), 6);

    case 6:
      return _ctx.chain(not_traced(_ctx.$$.f.v0), 7);

    case 7:
      return _ctx.pure();

    case 0:
      return _ctx.pure(p);

    case 1:
      throw p;

    default:
      throw new Error(\\"invalid state\\");
  }
}

function f_1(s, p) {
  var ctx = this;

  switch (s) {
    case 5:
      return ctx.chain(traced(1), 6);

    case 6:
      ctx.$._ = p;
      return ctx.chain(traced(2), 7);

    case 7:
      return ctx.pure(ctx.$._ + p);

    case 0:
      return ctx.pure(p);

    case 1:
      throw p;

    default:
      throw new Error(\\"invalid state\\");
  }
}

function func_1(s, p) {
  var f;
  var ctx = this;

  switch (s) {
    case 5:
      return ctx.chain(M.brk(\\"statement\\", \\"36:4-36:44\\"), 6);

    case 6:
      return ctx.chain(not_traced(f = M.fun(() => {
        var ctx = M.instance(f);
        return ctx.scope(5);
      }, meta$f2, f_1)), 7);

    case 7:
      return ctx.pure();

    case 0:
      return ctx.pure(p);

    case 1:
      throw p;

    default:
      throw new Error(\\"invalid state\\");
  }
}

function fn1_1(s, p) {
  var _fn1 = this,
      a,
      b,
      c;

  switch (s) {
    case 5:
      return _fn1.chain(M.brk(\\"statement\\", \\"41:2-41:16\\"), 6);

    case 6:
      _fn1.$.v5 = 10;
      return _fn1.chain(M.brk(\\"statement\\", \\"42:2-42:55\\"), 7);

    case 7:
      return _fn1.chain(_fn1.$$.f.mod.f2(), 8);

    case 8:
      _fn1.$._ = p;
      c = _fn1.$$.f.C;
      return _fn1.chain(new c(_fn1.$$.f.v0, _fn1.$.v5), 9);

    case 9:
      c = M.constr(p, \\"42:37-42:50\\");
      b = M.constr([c], \\"42:36-42:51\\");
      a = M.constr({
        a: _fn1.$._,
        b: b
      }, \\"42:18-42:53\\");
      return _fn1.chain(_fn1.$$.f.mod.some(a), 10);

    case 10:
      return _fn1.pure(p);

    case 0:
      return _fn1.pure(p);

    case 1:
      throw p;

    default:
      throw new Error(\\"invalid state\\");
  }
}

function objFn_1(s, p) {
  var _ctx = this,
      a;

  switch (s) {
    case 5:
      return _ctx.chain(M.brk(\\"statement\\", \\"47:4-47:31\\"), 6);

    case 6:
      a = _ctx.$$.f.Cl;
      return _ctx.chain(new a(), 7);

    case 7:
      a = M.constr(p, \\"47:12-47:20\\");
      return _ctx.chain(a.fn(_ctx.$.a, _ctx.$$.f.v0), 8);

    case 8:
      a = p;
      return _ctx.pure();

    case 0:
      return _ctx.pure(p);

    case 1:
      throw p;

    default:
      throw new Error(\\"invalid state\\");
  }
}

function _f_1(s, p) {
  var obj1,
      ctx = this,
      a;

  switch (s) {
    case 5:
      ctx.$.module.exports = ctx.$.exports;
      return ctx.chain(M.brk(\\"statement\\", \\"1:0-1:31\\"), 6);

    case 6:
      return ctx.chain(require(\\"mod\\"), 7);

    case 7:
      ({
        mod: ctx.$.mod
      } = p);
      return ctx.chain(M.brk(\\"statement\\", \\"3:0-3:12\\"), 8);

    case 8:
      ctx.$.v0 = 10;
      return ctx.chain(M.brk(\\"statement\\", \\"40:0-43:2\\"), 9);

    case 9:
      ctx.$.exports.fn2 = M.fun(function fn1() {
        var _fn1 = M.instance(fn1);

        _fn1.$$.f = ctx.$;
        return _fn1.scope(5);
      }, meta$fn1, fn1_1);
      return ctx.chain(M.brk(\\"statement\\", \\"45:0-49:2\\"), 10);

    case 10:
      obj1 = M.constr({
        objFn: M.fun(function objFn(a) {
          var _ctx = M.instance(objFn);

          _ctx.$$.f = ctx.$;
          _ctx.$.a = a;
          return _ctx.scope(5);
        }, meta$objFn, objFn_1)
      }, \\"45:13-49:1\\");
      a = ctx.$.module.exports;
      return ctx.pure(a);

    case 0:
      return ctx.pure(p);

    case 1:
      throw p;

    default:
      throw new Error(\\"invalid state\\");
  }
}"
`;

exports[`default transpiler 2`] = `
"\\"use strict\\";

var M = require(\\"../main\\");

var $module = M.module(\\"/__fixtures__/mod1.js\\"),
    meta$f = M.meta($module, {}),
    meta$run = M.meta($module, {});
module.exports = M.fun(function f(module, exports) {
  var ctx = M.instance(f);
  ctx.$.run = M.fun(function _run(i = 1) {
    var run = M.instance(_run);
    run.$$.f = ctx.$;
    run.$.i = i;
    return run.scope(5);
  }, meta$run, run_1);
  ctx.$.module = module;
  ctx.$.exports = exports;
  return ctx.scope(5);
}, meta$f, f_1)({}, {});

function run_1(s, p) {
  var run = this,
      a;

  switch (s) {
    case 5:
      return run.chain(M.brk(\\"statement\\", \\"6:2-6:9\\"), 6);

    case 6:
      run.$$.f.j += run.$.i;
      return run.chain(M.brk(\\"statement\\", \\"7:2-7:23\\"), 7);

    case 7:
      a = run.$.i++;
      return run.chain(run.$$.f.M2.f1(a), 8);

    case 8:
      return run.chain(run.$$.f.M2.f2(p, run.$.i), 9);

    case 9:
      return run.chain(M.brk(\\"statement\\", \\"8:2-8:9\\"), 10);

    case 10:
      run.$$.f.j += run.$.i;
      return run.chain(M.brk(\\"statement\\", \\"9:2-9:31\\"), 11);

    case 11:
      return run.chain(console.log(\`i:\${run.$.i},j:\${run.$$.f.j}\`), 12);

    case 12:
      return run.pure();

    case 0:
      return run.pure(p);

    case 1:
      throw p;

    default:
      throw new Error(\\"invalid state\\");
  }
}

function f_1(s, p) {
  var ctx = this,
      a;

  switch (s) {
    case 5:
      ctx.$.module.exports = ctx.$.exports;
      return ctx.chain(M.brk(\\"statement\\", \\"1:0-1:29\\"), 6);

    case 6:
      return ctx.chain(require(\\"./mod2\\"), 7);

    case 7:
      ctx.$.M2 = p;
      return ctx.chain(M.brk(\\"statement\\", \\"3:0-3:10\\"), 8);

    case 8:
      ctx.$.j = 0;
      return ctx.chain(M.brk(\\"statement\\", \\"12:0-12:6\\"), 9);

    case 9:
      return ctx.chain(ctx.$.run(), 10);

    case 10:
      return ctx.chain(M.brk(\\"statement\\", \\"13:0-13:6\\"), 11);

    case 11:
      return ctx.chain(ctx.$.run(), 12);

    case 12:
      a = ctx.$.module.exports;
      return ctx.pure(a);

    case 0:
      return ctx.pure(p);

    case 1:
      throw p;

    default:
      throw new Error(\\"invalid state\\");
  }
}"
`;

exports[`default transpiler 3`] = `
"\\"use strict\\";

var M = require(\\"../main\\");

var $module = M.module(\\"/__fixtures__/mod2.js\\"),
    meta$f = M.meta($module, {}),
    meta$f2 = M.meta($module, {}),
    meta$f1 = M.meta($module, {});
module.exports = M.fun(function f(module, exports) {
  var ctx = M.instance(f);
  ctx.$.f1 = M.fun(function _f1(i) {
    var f1 = M.instance(_f1);
    f1.$$.f = ctx.$;
    f1.$.i = i;
    return f1.scope(5);
  }, meta$f1, f1_1);
  ctx.$.f2 = M.fun(function _f2(i) {
    var f2 = M.instance(_f2);
    f2.$$.f = ctx.$;
    f2.$.i = i;
    return f2.scope(5);
  }, meta$f2, f2_1);
  ctx.$.module = module;
  ctx.$.exports = exports;
  return ctx.scope(5);
}, meta$f, f_1)({}, {});

function f1_1(s, p) {
  var f1 = this;

  switch (s) {
    case 5:
      return f1.chain(M.brk(\\"statement\\", \\"4:2-4:8\\"), 6);

    case 6:
      f1.$$.f.cnt++;
      return f1.chain(M.brk(\\"statement\\", \\"5:2-5:32\\"), 7);

    case 7:
      return f1.chain(console.log(\`f1:\${f1.$.i}@\${f1.$$.f.cnt}\`), 8);

    case 8:
      return f1.chain(M.brk(\\"statement\\", \\"6:2-6:13\\"), 9);

    case 9:
      return f1.pure(f1.$$.f.cnt);

    case 0:
      return f1.pure(p);

    case 1:
      throw p;

    default:
      throw new Error(\\"invalid state\\");
  }
}

function f2_1(s, p) {
  var f2 = this;

  switch (s) {
    case 5:
      return f2.chain(M.brk(\\"statement\\", \\"10:2-10:8\\"), 6);

    case 6:
      f2.$$.f.cnt++;
      return f2.chain(M.brk(\\"statement\\", \\"11:2-11:32\\"), 7);

    case 7:
      return f2.chain(console.log(\`f2:\${f2.$.i}@\${f2.$$.f.cnt}\`), 8);

    case 8:
      return f2.pure();

    case 0:
      return f2.pure(p);

    case 1:
      throw p;

    default:
      throw new Error(\\"invalid state\\");
  }
}

function f_1(s, p) {
  var ctx = this,
      a;

  switch (s) {
    case 5:
      ctx.$.module.exports = ctx.$.exports;
      ctx.$.exports.__esModule = true;
      ctx.$.exports.f1 = ctx.$.f1;
      ctx.$.exports.f2 = ctx.$.f2;
      return ctx.chain(M.brk(\\"statement\\", \\"1:0-1:12\\"), 6);

    case 6:
      ctx.$.cnt = 0;
      a = ctx.$.module.exports;
      return ctx.pure(a);

    case 0:
      return ctx.pure(p);

    case 1:
      throw p;

    default:
      throw new Error(\\"invalid state\\");
  }
}"
`;
